### DISCUSSION TITLE
Simple task retrying in the initial version?

### DISCUSSION THREAD

**Discussion by RealOrangeOne** [RealOrangeOne/django-tasks#47] [General]
Following some discussions at Djangocon Europe last week, it was mentioned that the lack of retry functionality, however basic, may be a non-starter for some users.

Adding relative simple retry logic ("If it fails, immediately try again `n` times") should be relatively simple to implement and expose, without adding much maintenance burden. But, the trade-off towards something which is more useful to more people (and thus aid adoption) might be worthwhile.

What do people think?

**Comment by jribbens** [comment]
I think at a minimum you need to be able to specify which exceptions cause a specific task to retry. e.g. a database exception means "retry", a `ValueError` means do not retry. Also not just how many times to re-try but how often. Perhaps `retry_on=(OperationalError, OSError, MemoryError), retry=(1, 10, 30, 60)` might mean "if you get a temporary database error, or an i/o error, or out of memory, retry 4 times, once after 1 second, once after 10 seconds, etc"? For the encouragement of fun yak shaving, you could also specify useful defaults for the value of `retry_on`.

**Reply by RealOrangeOne** [reply]
Retrying on a given exception is a very different feature to a simple "If this failed, retry".

I get that it's probably useful as a way to indicate whether to retry or not, but conditional retry is definitely far from "simple".

**Reply by jribbens** [reply]
I don't see why a tuple of expressions is not simple? And it's pretty vital, because otherwise you can end up spamming tasks that have no hope of succeeding.

**Comment by mgax** [comment]
It might be easier, at least initially, to encourage users to implement their own retry logic, and maybe provide some helpers.
- Make it possible for tasks to receive their `Task` object, similar to Celery's [bound tasks](https://docs.celeryq.dev/en/latest/userguide/tasks.html#bound-tasks). This way an exception handler could enqueue the task object again.
- Dispatch [signals](https://docs.djangoproject.com/en/5.0/topics/signals/) on task lifecycle events. A user could listen for a task-failed signal and queue it again.
- Document best practices for task retrying. Include an example with a retry counter and/or exponential backoff.
